File: go/Assembler/template_performance/performance_utils.go
// PerformanceUtils implementation for Go
// This module matches the C#/Rust structure

package template_performance

import (
	"assembler/template_engine"
	"assembler/template_loader"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"
)

type PerfSummaryRow struct {
	AppSite             string
	AppFile             string
	AppView             string
	Iterations          int
	NormalTimeNanos     int64
	PreProcessTimeNanos int64
	OutputSize          int
	ResultsMatch        string
	PerfDifference      string
}

// Helper methods for display
func (p *PerfSummaryRow) NormalTimeMs() float64 {
	return float64(p.NormalTimeNanos) / 1_000_000.0
}

func (p *PerfSummaryRow) PreProcessTimeMs() float64 {
	return float64(p.PreProcessTimeNanos) / 1_000_000.0
}

func RunPerformanceComparison(assemblerWebDir string, appSiteFilter string, skipDetails bool, enableJsonProcessing bool) []PerfSummaryRow {
	skipDetailsFlag := skipDetails
	iterations := 1000
	appSitesPath := filepath.Join(assemblerWebDir, "AppSites")
	entries, err := os.ReadDir(appSitesPath)
	if err != nil {
		fmt.Printf("❌ AppSites directory not found: %s\n", appSitesPath)
		return []PerfSummaryRow{}
	}
	var perfSummaryRows []PerfSummaryRow
	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}
		testAppSite := entry.Name()
		if appSiteFilter != "" && !strings.EqualFold(testAppSite, appSiteFilter) {
			continue
		}
		appSiteDir := filepath.Join(appSitesPath, testAppSite)
		htmlFiles, _ := os.ReadDir(appSiteDir)
		for _, htmlFile := range htmlFiles {
			if htmlFile.IsDir() || !strings.HasSuffix(htmlFile.Name(), ".html") {
				continue
			}
			appFileName := strings.TrimSuffix(htmlFile.Name(), ".html")
			templates := template_loader.LoadGetTemplateFiles(assemblerWebDir, testAppSite)
			siteTemplates := template_loader.LoadProcessGetTemplateFiles(assemblerWebDir, testAppSite)
			if len(templates) == 0 {
				if !skipDetailsFlag {
					fmt.Printf("❌ No templates found for %s\n", testAppSite)
				}
				continue
			}
			mainTemplateKey := strings.ToLower(fmt.Sprintf("%s_%s", testAppSite, appFileName))
			if _, ok := templates[mainTemplateKey]; !ok {
				if !skipDetailsFlag {
					fmt.Printf("❌ No main template found for %s\n", mainTemplateKey)
				}
				continue
			}
			if !skipDetailsFlag {
				fmt.Printf("Template Key: %s\n", mainTemplateKey)
				fmt.Printf("Templates available: %d\n", len(templates))
			}
			appViewScenarios := []struct{ AppView, AppViewPrefix string }{{"", ""}}
			viewsPath := filepath.Join(appSiteDir, "Views")
			if stat, err := os.Stat(viewsPath); err == nil && stat.IsDir() {
				viewFiles, _ := os.ReadDir(viewsPath)
				for _, viewFile := range viewFiles {
					if viewFile.IsDir() || !strings.HasSuffix(viewFile.Name(), ".html") {
						continue
					}
					viewName := strings.TrimSuffix(viewFile.Name(), ".html")
					viewNameLower := strings.ToLower(viewName)
					if idx := strings.Index(viewNameLower, "content"); idx > 0 {
						viewPart := viewName[:idx]
						if viewPart != "" {
							// Capitalize first letter to match other languages (C#, Rust, Node)
							appView := strings.ToUpper(string(viewPart[0])) + viewPart[1:]
							appViewPrefix := appView
							if len(appView) > 6 {
								appViewPrefix = appView[:6]
							}
							appViewScenarios = append(appViewScenarios, struct{ AppView, AppViewPrefix string }{appView, appViewPrefix})
						}
					}
				}
			}
			for _, scenario := range appViewScenarios {
				if !skipDetailsFlag {
					fmt.Println(strings.Repeat("-", 60))
					fmt.Printf(">>> GO SCENARIO : '%s', '%s', '%s', '%s'\n", testAppSite, appFileName, scenario.AppView, scenario.AppViewPrefix)
					fmt.Printf("Iterations per test: %d\n", iterations)
				}
				template_loader.ClearCache()
				normalEngine := template_engine.NewEngineNormal(appFileName)
				normalEngine.SetAppViewPrefix(scenario.AppViewPrefix)

				// Warmup - run a few iterations first to ensure consistent performance
				for warmup := 0; warmup < 100; warmup++ {
					normalEngine.MergeTemplates(testAppSite, appFileName, scenario.AppView, templates, enableJsonProcessing)
				}

				start := time.Now()
				var resultNormal string
				for i := 0; i < iterations; i++ {
					resultNormal = normalEngine.MergeTemplates(testAppSite, appFileName, scenario.AppView, templates, enableJsonProcessing)
				}
				normalTimeNanos := time.Since(start).Nanoseconds()
				if !skipDetailsFlag {
					fmt.Printf("[Normal Engine] %d iterations: %.2fms\n", iterations, float64(normalTimeNanos)/1_000_000.0)
					fmt.Printf("[Normal Engine] Avg: %.3fms per op, Output size: %d chars\n", float64(normalTimeNanos)/float64(iterations)/1_000_000.0, len(resultNormal))
				}
				template_loader.ClearCache()
				preprocessEngine := template_engine.NewEnginePreProcess(appFileName)
				preprocessEngine.SetAppViewPrefix(scenario.AppViewPrefix)

				// Warmup for PreProcess engine
				for warmup := 0; warmup < 100; warmup++ {
					preprocessEngine.MergeTemplates(testAppSite, appFileName, scenario.AppView, siteTemplates.Templates, enableJsonProcessing)
				}

				start = time.Now()
				var resultPreprocess string
				for i := 0; i < iterations; i++ {
					resultPreprocess = preprocessEngine.MergeTemplates(testAppSite, appFileName, scenario.AppView, siteTemplates.Templates, enableJsonProcessing)
				}
				preprocessTimeNanos := time.Since(start).Nanoseconds()
				resultsMatch := resultNormal == resultPreprocess
				if !skipDetailsFlag {
					fmt.Printf("[PreProcess Engine] %d iterations: %.2fms\n", iterations, float64(preprocessTimeNanos)/1_000_000.0)
					fmt.Printf("[PreProcess Engine] Avg: %.3fms per op, Output size: %d chars\n", float64(preprocessTimeNanos)/float64(iterations)/1_000_000.0, len(resultPreprocess))
					fmt.Printf("Results match: %s\n", map[bool]string{true: "✅ YES", false: "❌ NO"}[resultsMatch])
				}
				perfDiff := "0%"
				if normalTimeNanos > 0 {
					perfDiff = fmt.Sprintf("%.1f%%", float64(preprocessTimeNanos-normalTimeNanos)/float64(normalTimeNanos)*100.0)
				}
				perfSummaryRows = append(perfSummaryRows, PerfSummaryRow{
					AppSite:             testAppSite,
					AppFile:             appFileName,
					AppView:             scenario.AppView,
					Iterations:          iterations,
					NormalTimeNanos:     normalTimeNanos,
					PreProcessTimeNanos: preprocessTimeNanos,
					OutputSize:          len(resultNormal),
					ResultsMatch:        map[bool]string{true: "YES", false: "NO"}[resultsMatch],
					PerfDifference:      perfDiff,
				})
			}
		}
	}
	return perfSummaryRows
}

func PrintPerfSummaryTable(assemblerWebDir string, perfSummaryRows []PerfSummaryRow) {
	if len(perfSummaryRows) == 0 {
		return
	}
	fmt.Println("\n==================== GO PERFORMANCE SUMMARY ====================")
	headers := []string{"AppSite", "AppView", "Normal(ms)", "PreProc(ms)", "Match", "PerfDiff"}
	colCount := len(headers)
	widths := make([]int, colCount)
	for i, h := range headers {
		widths[i] = len(h)
	}
	for _, row := range perfSummaryRows {
		widths[0] = maxInt(widths[0], len(row.AppSite))
		widths[1] = maxInt(widths[1], len(row.AppView))
		widths[2] = maxInt(widths[2], len(fmt.Sprintf("%.2f", row.NormalTimeMs())))
		widths[3] = maxInt(widths[3], len(fmt.Sprintf("%.2f", row.PreProcessTimeMs())))
		widths[4] = maxInt(widths[4], len(row.ResultsMatch))
		widths[5] = maxInt(widths[5], len(row.PerfDifference))
	}

	// Print header
	fmt.Print("| ")
	for i, h := range headers {
		fmt.Printf("%-*s", widths[i], h)
		if i < colCount-1 {
			fmt.Print(" | ")
		}
	}
	fmt.Println(" |")
	// Print divider
	fmt.Print("|")
	for i := range headers {
		fmt.Printf(" %s ", strings.Repeat("-", widths[i]))
		if i < colCount-1 {
			fmt.Print("|")
		}
	}
	fmt.Println("|")
	// Print rows
	for _, row := range perfSummaryRows {
		fmt.Print("| ")
		fmt.Printf("%-*s", widths[0], row.AppSite)
		fmt.Print(" | ")
		fmt.Printf("%-*s", widths[1], row.AppView)
		fmt.Print(" | ")
		fmt.Printf("%-*.2f", widths[2], row.NormalTimeMs())
		fmt.Print(" | ")
		fmt.Printf("%-*.2f", widths[3], row.PreProcessTimeMs())
		fmt.Print(" | ")
		fmt.Printf("%-*s", widths[4], row.ResultsMatch)
		fmt.Print(" | ")
		fmt.Printf("%-*s", widths[5], row.PerfDifference)
		fmt.Println(" |")
	}
	fmt.Print("|")
	for i := range headers {
		fmt.Printf(" %s ", strings.Repeat("-", widths[i]))
		if i < colCount-1 {
			fmt.Print("|")
		}
	}
	fmt.Println("|")
	// Save performance summary to file
	if err := os.MkdirAll(assemblerWebDir, 0755); err != nil {
		fmt.Printf("❌ Error creating output directory: %v\n", err)
		return
	}
	perfJsonFile := filepath.Join(assemblerWebDir, "go_perfsummary.json")
	perfHtmlFile := filepath.Join(assemblerWebDir, "go_perfsummary.html")
	perfJson, _ := json.MarshalIndent(perfSummaryRows, "", "  ")
	if err := os.WriteFile(perfJsonFile, perfJson, 0644); err != nil {
		fmt.Printf("❌ Error writing performance JSON file: %v\n", err)
	} else {
		fmt.Printf("Performance summary JSON saved to: %s\n", perfJsonFile)
	}
	// Generate HTML performance summary table
	html := "<html><head><title>Go Performance Summary Table</title></head><body>\n"
	html += "<h1>Go Performance Summary</h1>\n"
	html += "<table border='1' cellpadding='5' cellspacing='0'>\n"
	html += "<tr><th>AppSite</th><th>AppFile</th><th>AppView</th><th>Iterations</th><th>NormalMs</th><th>PreProcessMs</th><th>OutputSize</th><th>Match</th><th>PerfDiff</th></tr>\n"
	for _, row := range perfSummaryRows {
		html += fmt.Sprintf("<tr><td>%s</td><td>%s</td><td>%s</td><td>%d</td><td>%.2f</td><td>%.2f</td><td>%d</td><td>%s</td><td>%s</td></tr>\n",
			row.AppSite, row.AppFile, row.AppView, row.Iterations,
			row.NormalTimeMs(), row.PreProcessTimeMs(),
			row.OutputSize, row.ResultsMatch, row.PerfDifference)
	}
	html += "</table>\n</body></html>"
	if err := os.WriteFile(perfHtmlFile, []byte(html), 0644); err != nil {
		fmt.Printf("❌ Error writing performance HTML file: %v\n", err)
	} else {
		fmt.Printf("Performance summary HTML saved to: %s\n", perfHtmlFile)
	}
}

func maxInt(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: rust/Assembler/src/template_performance/performance_utils.rs
use std::time::Instant;
use std::fs;
use serde::{Serialize, Deserialize};
use crate::template_loader::loader_normal::LoaderNormal;
use crate::template_loader::loader_preprocess::LoaderPreProcess;
use crate::template_engine::engine_normal::EngineNormal;
use crate::template_engine::engine_preprocess::EnginePreProcess;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PerfSummaryRow {
    pub app_site: String,
    pub app_file: String,
    pub app_view: String,
    pub iterations: i32,
    pub normal_time_nanos: u128,
    pub preprocess_time_nanos: u128,
    pub output_size: usize,
    pub results_match: String,
    pub perf_difference: String,
}

impl PerfSummaryRow {
    pub fn normal_time_ms(&self) -> f64 {
        self.normal_time_nanos as f64 / 1_000_000.0
    }
    
    pub fn preprocess_time_ms(&self) -> f64 {
        self.preprocess_time_nanos as f64 / 1_000_000.0
    }
}

pub struct PerformanceUtils;

impl PerformanceUtils 
{
    /// Runs performance comparison and returns summary rows
    pub fn run_performance_comparison(assembler_web_dir_path: &str, app_site_filter: Option<&str>, skip_details: bool, enable_json_processing: bool) -> Vec<PerfSummaryRow> 
	{
        let iterations = 1000;
        let app_sites_path = std::path::Path::new(assembler_web_dir_path).join("AppSites");
        if !app_sites_path.exists() {
            if !skip_details {
                println!("❌ AppSites directory not found: {:?}", app_sites_path);
            }
            return Vec::new();
        }
        
        let mut perf_summary_rows: Vec<PerfSummaryRow> = Vec::new();
        
        for entry in std::fs::read_dir(&app_sites_path).unwrap() {
            let entry = entry.unwrap();
            if !entry.path().is_dir() { continue; }
            
            let test_app_site = entry.file_name().to_string_lossy().to_string();
            if let Some(filter) = app_site_filter {
                if !test_app_site.eq_ignore_ascii_case(filter) {
                    continue;
                }
            }
            
            let app_site_dir = entry.path();
            let html_files: Vec<_> = std::fs::read_dir(&app_site_dir)
                .unwrap()
                .filter_map(|entry| {
                    let entry = entry.ok()?;
                    let path = entry.path();
                    if path.extension()?.to_str()? == "html" {
                        Some(path)
                    } else {
                        None
                    }
                })
                .collect();
                
            for html_file_path in html_files {
                let app_file_name = match html_file_path.file_stem() {
                    Some(stem) => stem.to_string_lossy().to_string(),
                    None => continue,
                };
                
                // Clear cache and load templates
                LoaderNormal::clear_cache();
                LoaderPreProcess::clear_cache();
                let mut templates = LoaderNormal::load_get_template_files(assembler_web_dir_path, &test_app_site);
                
                let site_templates = LoaderPreProcess::load_process_get_template_files(assembler_web_dir_path, &test_app_site);
                
                if templates.is_empty() {
                    if !skip_details {
                        println!("❌ No templates found for {}", test_app_site);
                    }
                    continue;
                }
                
                let main_template_key = format!("{}_{}", test_app_site, app_file_name).to_lowercase();
                if !templates.contains_key(&main_template_key) {
                    if !skip_details {
                        println!("❌ No main template found for {}", main_template_key);
                    }
                    continue;
                }
                
                if !skip_details {
                    println!("Template Key: {}", main_template_key);
                    println!("Templates available: {}", templates.len());
                }
                
                // Build AppView scenarios
                let mut app_view_scenarios = vec![(String::new(), String::new())]; // No AppView
                
                let views_path = app_site_dir.join("Views");
                if views_path.exists() {
                    for view_entry in std::fs::read_dir(&views_path).unwrap() {
                        let view_entry = view_entry.unwrap();
                        if let Some(view_name) = view_entry.path().file_stem() {
                            let view_name_str = view_name.to_string_lossy().to_string();
                            let view_name_lower = view_name_str.to_lowercase();
                            if view_name_lower.contains("content") {
                                if let Some(content_index) = view_name_lower.find("content") {
                                    if content_index > 0 {
                                        let view_part = &view_name_str[..content_index];
                                        if !view_part.is_empty() {
                                            let app_view = format!("{}{}", 
                                                view_part.chars().next().unwrap().to_uppercase(),
                                                &view_part[1..]
                                            );
                                            let app_view_prefix = app_view.chars().take(6).collect::<String>();
                                            app_view_scenarios.push((app_view, app_view_prefix));
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
                
                for (app_view, app_view_prefix) in app_view_scenarios {
                    if !skip_details {
                        println!("{}", "-".repeat(60));
                        println!(">>> RUST SCENARIO : '{}', '{}', '{}', '{}'", test_app_site, app_file_name, app_view, app_view_prefix);
                        println!("Iterations per test: {}", iterations);
                    }
                    
                    // Normal Engine
                    LoaderNormal::clear_cache();
                    LoaderPreProcess::clear_cache();
                    let mut normal_engine = EngineNormal::new(app_file_name.clone());
                    normal_engine.set_app_view_prefix(app_view_prefix.clone());
                    
                    // Warmup - run a few iterations first to ensure consistent performance
                    for _ in 0..100 {
                        normal_engine.merge_templates(&test_app_site, &app_file_name, 
                            if app_view.is_empty() { None } else { Some(&app_view) }, &mut templates, enable_json_processing);
                    }
                    
                    let start = Instant::now();
                    let mut result_normal = String::new();
                    for _ in 0..iterations {
                        result_normal = normal_engine.merge_templates(&test_app_site, &app_file_name, 
                            if app_view.is_empty() { None } else { Some(&app_view) }, &mut templates, enable_json_processing);
                    }
                    let normal_duration = start.elapsed();
                    let normal_time_nanos = normal_duration.as_nanos();
                    
                    if !skip_details {
                        println!("[Normal Engine] {} iterations: {:.2}ms", iterations, normal_time_nanos as f64 / 1_000_000.0);
                        println!("[Normal Engine] Avg: {:.3}ms per op, Output size: {} chars", 
                            normal_time_nanos as f64 / 1_000_000.0 / iterations as f64, result_normal.len());
                    }
                    
                    // PreProcess Engine
                    LoaderNormal::clear_cache();
                    LoaderPreProcess::clear_cache();
                    let mut preprocess_engine = EnginePreProcess::new(app_file_name.clone());
                    preprocess_engine.set_app_view_prefix(app_view_prefix.clone());
                    
                    // Warmup for PreProcess engine
                    for _ in 0..100 {
                        preprocess_engine.merge_templates(&test_app_site, &app_file_name,
                            if app_view.is_empty() { None } else { Some(&app_view) }, &site_templates.templates, enable_json_processing);
                    }
                    
                    let start = Instant::now();
                    let mut result_preprocess = String::new();
                    for _ in 0..iterations {
                        result_preprocess = preprocess_engine.merge_templates(&test_app_site, &app_file_name,
                            if app_view.is_empty() { None } else { Some(&app_view) }, &site_templates.templates, enable_json_processing);
                    }
                    let preprocess_duration = start.elapsed();
                    let preprocess_time_nanos = preprocess_duration.as_nanos();
                    
                    if !skip_details {
                        println!("[PreProcess Engine] {} iterations: {:.2}ms", iterations, preprocess_time_nanos as f64 / 1_000_000.0);
                        println!("[PreProcess Engine] Avg: {:.3}ms per op, Output size: {} chars",
                            preprocess_time_nanos as f64 / 1_000_000.0 / iterations as f64, result_preprocess.len());
                        
                        // Comparison
                        println!(">>> RUST PERFORMANCE COMPARISON:");
                        println!("{}", "-".repeat(50));
                        let difference_nanos = preprocess_time_nanos as i128 - normal_time_nanos as i128;
                        let difference_ms = difference_nanos as f64 / 1_000_000.0;
                        let difference_percent = if normal_time_nanos > 0 {
                            (difference_nanos as f64 / normal_time_nanos as f64) * 100.0
                        } else {
                            0.0
                        };
                        
                        println!("Time difference: {:.2}ms ({:.1}%)", difference_ms, difference_percent);
                        let results_match = result_normal == result_preprocess;
                        println!("Results match: {}", if results_match { "✅ YES" } else { "❌ NO" });
                    }
                    
                    let results_match = result_normal == result_preprocess;
                    let perf_diff = if normal_time_nanos > 0 {
                        format!("{:.1}%", (preprocess_time_nanos as f64 - normal_time_nanos as f64) / normal_time_nanos as f64 * 100.0)
                    } else {
                        "0%".to_string()
                    };
                    
                    perf_summary_rows.push(PerfSummaryRow {
                        app_site: test_app_site.clone(),
                        app_file: app_file_name.clone(),
                        app_view: app_view.clone(),
                        iterations,
                        normal_time_nanos,
                        preprocess_time_nanos,
                        output_size: result_normal.len(),
                        results_match: if results_match { "YES".to_string() } else { "NO".to_string() },
                        perf_difference: perf_diff,
                    });
                }
            }
        }
        
        perf_summary_rows
    }

    /// Prints the performance summary table in markdown format
    pub fn print_perf_summary_table(assembler_web_dir_path: &str, summary_rows: &Vec<PerfSummaryRow>) 
	{
        if summary_rows.is_empty() {
            return;
        }
        
        println!("\n==================== RUST PERFORMANCE SUMMARY ====================\n");
        
        let headers = vec!["AppSite", "AppView", "Normal(ms)", "PreProc(ms)", "Match", "PerfDiff"];
        let col_count = headers.len();
        let mut widths = vec![0; col_count];
        
        // Calculate column widths
        for i in 0..col_count {
            widths[i] = headers[i].len();
        }
        
        for row in summary_rows {
            widths[0] = widths[0].max(row.app_site.len());
            widths[1] = widths[1].max(row.app_view.len());
            widths[2] = widths[2].max(format!("{:.2}", row.normal_time_ms()).len());
            widths[3] = widths[3].max(format!("{:.2}", row.preprocess_time_ms()).len());
            widths[4] = widths[4].max(row.results_match.len());
            widths[5] = widths[5].max(row.perf_difference.len());
        }
        
        // Print header
        print!("| ");
        for i in 0..col_count {
            print!("{}", format!("{:<width$}", headers[i], width = widths[i]));
            if i < col_count - 1 {
                print!(" | ");
            }
        }
        println!(" |");
        
        // Print divider
        print!("|");
        for i in 0..col_count {
            print!(" {} ", "-".repeat(widths[i]));
            if i < col_count - 1 {
                print!("|");
            }
        }
        println!("|");
        
        // Print rows
        for row in summary_rows {
            print!("| ");
            print!("{:<width$}", row.app_site, width = widths[0]);
            print!(" | ");
            print!("{:<width$}", row.app_view, width = widths[1]);
            print!(" | ");
            print!("{:<width$.2}", format!("{:.2}", row.normal_time_ms()), width = widths[2]);
            print!(" | ");
            print!("{:<width$.2}", format!("{:.2}", row.preprocess_time_ms()), width = widths[3]);
            print!(" | ");
            print!("{:<width$}", row.results_match, width = widths[4]);
            print!(" | ");
            print!("{:<width$}", row.perf_difference, width = widths[5]);
            println!(" |");
        }
        print!("|");
        for i in 0..col_count {
            print!(" {} ", "-".repeat(widths[i]));
            if i < col_count - 1 {
                print!("|");
            }
        }
        println!("|");
        
        // Save performance summary to file
        let output_dir = assembler_web_dir_path;
        if let Err(e) = std::fs::create_dir_all(&output_dir) {
            println!("❌ Error creating output directory: {}", e);
            return;
        }
        
        let perf_json_file = format!("{}/rust_perfsummary.json", output_dir);
        let perf_html_file = format!("{}/rust_perfsummary.html", output_dir);
        
        let perf_json = serde_json::to_string_pretty(&summary_rows).unwrap();
        if let Err(e) = fs::write(&perf_json_file, perf_json) {
            println!("❌ Error writing performance JSON file: {}", e);
        } else {
            println!("Performance summary JSON saved to: {}", perf_json_file);
        }
        
        // Generate HTML performance summary table
        let mut html = String::new();
        html.push_str("<html><head><title>Rust Performance Summary Table</title><style>table{border-collapse:collapse;}th,td{border:1px solid #888;padding:4px;}th{background:#eee;}</style></head><body>\n");
        html.push_str("<h2>Rust Performance Summary Table</h2>\n");
        html.push_str("<table>\n");
        html.push_str("<tr><th>AppSite</th><th>AppView</th><th>Normal(ms)</th><th>PreProc(ms)</th><th>Match</th><th>PerfDiff</th></tr>\n");
        
        for row in summary_rows {
            html.push_str(&format!(
                "<tr><td>{}</td><td>{}</td><td>{:.2}</td><td>{:.2}</td><td>{}</td><td>{}</td></tr>\n",
                row.app_site, row.app_view,
                row.normal_time_ms(), row.preprocess_time_ms(),
                row.results_match, row.perf_difference
            ));
        }
        
        html.push_str("</table>\n</body></html>");
        if let Err(e) = fs::write(&perf_html_file, html) {
            println!("❌ Error writing performance HTML file: {}", e);
        } else {
            println!("Performance summary HTML saved to: {}", perf_html_file);
        }
    }
}
